From 83643096200800bbaf605d7800e8e5a490a67e31 Mon Sep 17 00:00:00 2001
From: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
Date: Fri, 15 Aug 2025 17:39:01 -0700
Subject: [PATCH 117/128] mtk: mac80211: Add new struct for TWT teardown flow

Introduce struct ieee80211_twt_teardown. The old code use a u8 to represent id
which can't represent all the scenarios when negotiation type > 0 or
when all != 0

Signed-off-by: Yi-Chia Hsieh <Yi-Chia.Hsieh@mediatek.com>
---
 include/linux/ieee80211-he.h |  6 ++++++
 include/net/mac80211.h       |  3 ++-
 net/mac80211/driver-ops.h    |  6 +++---
 net/mac80211/s1g.c           | 18 ++++++++++--------
 net/mac80211/trace.h         |  7 ++++---
 5 files changed, 25 insertions(+), 15 deletions(-)

diff --git a/include/linux/ieee80211-he.h b/include/linux/ieee80211-he.h
index b652ede7..879da4eb 100644
--- a/include/linux/ieee80211-he.h
+++ b/include/linux/ieee80211-he.h
@@ -70,6 +70,12 @@ struct ieee80211_twt_setup {
 	u8 params[];
 } __packed;
 
+struct ieee80211_twt_teardown {
+	u8 id:5;
+	u8 neg_type:2;
+	u8 all:1;
+} __packed;
+
 /**
  * struct ieee80211_he_cap_elem - HE capabilities element
  * @mac_cap_info: HE MAC Capabilities Information
diff --git a/include/net/mac80211.h b/include/net/mac80211.h
index 583fbc1c..44937671 100644
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -4929,7 +4929,8 @@ struct ieee80211_ops {
 			      struct ieee80211_sta *sta,
 			      struct ieee80211_twt_setup *twt);
 	void (*twt_teardown_request)(struct ieee80211_hw *hw,
-				     struct ieee80211_sta *sta, u8 flowid);
+				     struct ieee80211_sta *sta,
+				     struct ieee80211_twt_teardown *td);
 	int (*add_btwt)(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 			struct cfg80211_broadcast_twt_settings *btwt);
 	int (*del_btwt)(struct ieee80211_hw *hw, struct ieee80211_vif *vif, u8 id);
diff --git a/net/mac80211/driver-ops.h b/net/mac80211/driver-ops.h
index e8e48a2b..0c9385e8 100644
--- a/net/mac80211/driver-ops.h
+++ b/net/mac80211/driver-ops.h
@@ -1657,7 +1657,7 @@ static inline void drv_add_twt_setup(struct ieee80211_local *local,
 static inline void drv_twt_teardown_request(struct ieee80211_local *local,
 					    struct ieee80211_sub_if_data *sdata,
 					    struct ieee80211_sta *sta,
-					    u8 flowid)
+					    struct ieee80211_twt_teardown *td)
 {
 	might_sleep();
 	lockdep_assert_wiphy(local->hw.wiphy);
@@ -1667,8 +1667,8 @@ static inline void drv_twt_teardown_request(struct ieee80211_local *local,
 	if (!local->ops->twt_teardown_request)
 		return;
 
-	trace_drv_twt_teardown_request(local, sta, flowid);
-	local->ops->twt_teardown_request(&local->hw, sta, flowid);
+	trace_drv_twt_teardown_request(local, sta, td);
+	local->ops->twt_teardown_request(&local->hw, sta, td);
 	trace_drv_return_void(local);
 }
 
diff --git a/net/mac80211/s1g.c b/net/mac80211/s1g.c
index f61a7957..01388407 100644
--- a/net/mac80211/s1g.c
+++ b/net/mac80211/s1g.c
@@ -63,7 +63,8 @@ ieee80211_s1g_send_twt_setup(struct ieee80211_sub_if_data *sdata, const u8 *da,
 
 static void
 ieee80211_s1g_send_twt_teardown(struct ieee80211_sub_if_data *sdata,
-				const u8 *da, const u8 *bssid, u8 flowid)
+				const u8 *da, const u8 *bssid,
+				struct ieee80211_twt_teardown *td)
 {
 	struct ieee80211_local *local = sdata->local;
 	struct ieee80211_mgmt *mgmt;
@@ -86,7 +87,7 @@ ieee80211_s1g_send_twt_teardown(struct ieee80211_sub_if_data *sdata,
 	mgmt->u.action.category = WLAN_CATEGORY_S1G;
 	mgmt->u.action.u.s1g.action_code = WLAN_S1G_TWT_TEARDOWN;
 	id = (u8 *)mgmt->u.action.u.s1g.variable;
-	*id = flowid;
+	*id = td->id;
 
 	IEEE80211_SKB_CB(skb)->flags |= IEEE80211_TX_INTFL_DONT_ENCRYPT |
 					IEEE80211_TX_CTL_REQ_TX_STATUS;
@@ -129,7 +130,7 @@ ieee80211_s1g_rx_twt_teardown(struct ieee80211_sub_if_data *sdata,
 	struct ieee80211_mgmt *mgmt = (struct ieee80211_mgmt *)skb->data;
 
 	drv_twt_teardown_request(sdata->local, sdata, &sta->sta,
-				 mgmt->u.action.u.s1g.variable[0]);
+		(struct ieee80211_twt_teardown *) mgmt->u.action.u.s1g.variable);
 }
 
 static void
@@ -139,13 +140,14 @@ ieee80211_s1g_tx_twt_setup_fail(struct ieee80211_sub_if_data *sdata,
 	struct ieee80211_mgmt *mgmt = (struct ieee80211_mgmt *)skb->data;
 	struct ieee80211_twt_setup *twt = (void *)mgmt->u.action.u.s1g.variable;
 	struct ieee80211_twt_params *twt_agrt = (void *)twt->params;
-	u8 flowid = le16_get_bits(twt_agrt->req_type,
-				  IEEE80211_TWT_REQTYPE_FLOWID);
+	struct ieee80211_twt_teardown td;
 
-	drv_twt_teardown_request(sdata->local, sdata, &sta->sta, flowid);
+	td.id = le16_get_bits(twt_agrt->req_type,
+			      IEEE80211_TWT_REQTYPE_FLOWID);
 
-	ieee80211_s1g_send_twt_teardown(sdata, mgmt->sa, sdata->vif.addr,
-					flowid);
+	drv_twt_teardown_request(sdata->local, sdata, &sta->sta, &td);
+
+	ieee80211_s1g_send_twt_teardown(sdata, mgmt->sa, sdata->vif.addr, &td);
 }
 
 void ieee80211_s1g_rx_twt_action(struct ieee80211_sub_if_data *sdata,
diff --git a/net/mac80211/trace.h b/net/mac80211/trace.h
index 1373cb0d..92a72920 100644
--- a/net/mac80211/trace.h
+++ b/net/mac80211/trace.h
@@ -2530,9 +2530,10 @@ TRACE_EVENT(drv_add_twt_setup,
 
 TRACE_EVENT(drv_twt_teardown_request,
 	TP_PROTO(struct ieee80211_local *local,
-		 struct ieee80211_sta *sta, u8 flowid),
+		 struct ieee80211_sta *sta,
+		 struct ieee80211_twt_teardown *td),
 
-	TP_ARGS(local, sta, flowid),
+	TP_ARGS(local, sta, td),
 
 	TP_STRUCT__entry(
 		LOCAL_ENTRY
@@ -2543,7 +2544,7 @@ TRACE_EVENT(drv_twt_teardown_request,
 	TP_fast_assign(
 		LOCAL_ASSIGN;
 		STA_ASSIGN;
-		__entry->flowid = flowid;
+		__entry->flowid = td->id;
 	),
 
 	TP_printk(
-- 
2.45.2

